test/passes/optimize-instructions_all-features.txt

```diff
@@ -3713,7 +3713,11 @@
   )
  )
  (func $if-arms-subtype (result anyref)
-  (ref.null extern)
+  (if (result anyref)
+   (i32.const 0)
+   (ref.null extern)
+   (ref.null func)
```

Comment:
Currently this is identical to the original function in optimize-instruction_all-features.wast, so it does not serve any testing purpose.

I suggest to change `if-arms-subtype` function in optimize-instructions_all-feature.wast to this:
```wat
(func $if-arms-subtype (result anyref)
  (if (result anyref)
    (i32.const 0)
    (ref.null extern)
    (ref.null extern)
  )
)
```

---

Background:
This test was first added in #2451 so that when `if`'s true body and false body are identical, it is possible to fold the expression to if's true body even if the `if`'s true body's type is a subtype of `if`'s type. The relevant optimizing routine is [here](https://github.com/WebAssembly/binaryen/blob/0c53fb5d84afe32bfdf13231908aad760cb1d89d/src/passes/OptimizeInstructions.cpp#L559-L590).

So, before #2451, only this was able to be folded:
```wat
(if (result i32)
  (i32.const 0)
  (i32.const 3)
  (i32.const 3)
)
```
Note that `if`'s true body (and also false body)'s type `i32` is the same as the `if`'s type.

#2451 allowed this to be folded too:
```wat
(if (result anyref)
  (i32.const 0)
  (ref.null) ;; nullref type
  (ref.null) ;; nullref type
)
```
Then `ref.null` had `nullref` type, which used to be a subtype of `anyref`. So now we were able to fold even if `if`'s bodies are subtype (`nullref`) of `if`'s type (`anyref`).

Note that this optimization only happens when `if`'s true body and false body are [identical](https://github.com/WebAssembly/binaryen/blob/0c53fb5d84afe32bfdf13231908aad760cb1d89d/src/passes/OptimizeInstructions.cpp#L560). But it looks like #3084 mistakenly changed this code and made these to bodies not identical:
```wat
(if (result anyref)
  (i32.const 0)
  (ref.null extern)
  (ref.null func)
)
```
So now these bodies are not identical, this should not have been optimized. But this was still optimized and went unnoticed because of the bug you fixed in this PR.

So, long story short, we should make them identical again. :)

Reply:
And to make it more readable, it might be better to add a short comment for `$if-arms-subtype` in optimize-instruction_all-features.wast, something like
```
This function tests if 'if' can be folded as its true body when
1. It's true and false body are identical
2. When if's true (and false) body's type is not identical but a subtype of 'if''s type
```

Reply:
Maybe in addition to the case you suggested, we can also keep the current one that is the same in the wast and the output as a negative test showing that the arms are correctly not merged.